backend/routes/providers.js

```diff
@@ -6,25 +6,86 @@ const router = require('express').Router()
 router.get('/providers', async (req, res) => {
     const { search } = req.query
 
+    if(!search || search.trim() === '') {
+        return res.status(400).json({ error: 'No search query inputted' })
+    }
+
     try {
         const providers = await prisma.user.findMany({
-            where: {
-                role: 'PROVIDER',
-                name: { contains: search, mode: 'insensitive' }
-            },
+            where: { role: 'PROVIDER' },
             select: {
                 id: true,
                 name: true,
                 servicesOffered: true
             }
         })
-        res.json(providers)
+
+        const providersScored = providers.map(provider => {
+            const similarityScore = findSimilarityScore(search.toLowerCase(), provider.name.toLowerCase())
+            return {...provider, similarityScore}
+        })
+
+        const bestMatchedProviders = providersScored
+            .filter(provider => provider.similarityScore >= 0.55) // ignore names with scores below 0.55
+            .sort((a, b) => b.similarityScore - a.similarityScore)
+
+        res.json(bestMatchedProviders)
     } catch (error) {
         console.log(error);
         return res.status(500).json({ error: 'Server error' })  
     }
 }) 
 
+// Fuzzy Search function
+function findSimilarityScore(input, target) {
+    let i = 0
+    let j = 0
+    let prefixLength = 0
+
+    while(i < input.length && i < target.length && input[i] === target[i] && prefixLength < 4) {
+        prefixLength++
+        i++
+    }
+
+    let bonus = prefixLength * 0.05
```

Comment:
Nit: Put 0.05 into a global constant eg. (`BONUS_MULTIPLIER`)

This makes it so it can be reused later (though I don't expect that to be needed in your case) and only needs to be modified in one place
```suggestion
    let bonus = prefixLength * BONUS_MULTIPLIER
```